c_crypto.c

```diff
@@ -166,120 +174,365 @@ unsigned char *digest_message_SHA_256(unsigned char *message,
+            EVP_CIPHER_CTX_free(ctx);
+            print_last_error("EVP_CIPHER_CTX_ctrl (SET_TAG) failed");
+            return 1;
+        }
+
+        encrypted_length -=
+            AES_GCM_TAG_SIZE;  // Adjust the encrypted length to exclude the tag
+    }
+
     if (!EVP_DecryptUpdate(ctx, ret, (int *)ret_length, encrypted,
                            encrypted_length)) {
         EVP_CIPHER_CTX_free(ctx);
         print_last_error("EVP_DecryptUpdate failed");
+        return 1;
     }
+
     unsigned int temp_len;
     if (!EVP_DecryptFinal_ex(ctx, ret + *ret_length, (int *)&temp_len)) {
         EVP_CIPHER_CTX_free(ctx);
         print_last_error("EVP_DecryptFinal_ex failed");
+        return 1;
     }
     *ret_length += temp_len;
+
     EVP_CIPHER_CTX_free(ctx);
+    return 0;
 }
 
-unsigned char *symmetric_encrypt_authenticate(
+unsigned int get_expected_encrypted_total_length(unsigned int buf_length,
+                                                 unsigned int iv_size,
+                                                 unsigned int mac_key_size,
+                                                 char enc_mode,
+                                                 char no_hmac_mode) {
+    unsigned int encrypted_total_length = 0;
+    if (enc_mode == AES_128_CBC) {
+        // This requires, paddings, making the encrypted length multiples of the
+        // block size (key size)
+        encrypted_total_length = ((buf_length / iv_size) + 1) * iv_size;
+    } else if (enc_mode == AES_128_CTR) {
+        // The encrypted length is same on CTR mode.
+        encrypted_total_length = buf_length;
+    } else if (enc_mode == AES_128_GCM) {
+        encrypted_total_length =
+            buf_length +
+            AES_GCM_TAG_SIZE;  // GCM_TAG //TODO: Check. Tag size default is 12.
+    }
+    if (no_hmac_mode == 0) {
+        encrypted_total_length =
+            iv_size + encrypted_total_length + mac_key_size;
+    } else {
+        encrypted_total_length = iv_size + encrypted_total_length;
+    }
+    return encrypted_total_length;
+}
+
+static int get_symmetric_encrypt_authenticate_buffer(
+    unsigned char *buf, unsigned int buf_length, unsigned char *mac_key,
+    unsigned int mac_key_size, unsigned char *cipher_key,
+    unsigned int cipher_key_size, unsigned int iv_size, char enc_mode,
+    char no_hmac_mode, unsigned int expected_encrypted_total_length,
+    unsigned char *ret, unsigned int *ret_length) {
+    // The return of encrypt_AES will look like this.
+    // ret = IV (16) + encrypted(IV+buf) + (optional) HMAC(IV + encrypted) (32)
+    // First attach IV.
+    generate_nonce(iv_size, ret);
+    unsigned int total_length = 0;
+    // Attach encrypted buffer
+    if (cipher_key_size == AES_128_KEY_SIZE_IN_BYTES) {
+        if (encrypt_AES(buf, buf_length, cipher_key, ret, enc_mode,
+                        ret + iv_size, &total_length)) {
+            printf("AES encryption failed!\n");
+            return 1;
```

Comment:
Please use negative number.

Reply:
Same as above in eb82041512f68b8ba6cb566fb3615a80f8d0c0c5.